snapcraft/tests/unit/plugins/test_kernel.py

```diff
@@ -856,7 +859,8 @@ def test_build_with_kconfigfile_and_no_firmware(self):
                                      self.project_options)
 
         self._simulate_build(
-            plugin.sourcedir, plugin.builddir, plugin.installdir, do_dtbs=True)
+            plugin.sourcedir, plugin.builddir, plugin.installdir, do_dtbs=True,
+            do_firmware=False)
```

Comment:
I can't claim complete familiarity with this test, but it doesn't appear that anything verifies that the firmware is or isn't moved. Is that true?

Reply:
No, that test was actually bogus: it tested the 'kernel-with-firmware: false' condition but it was passing a build environment that actually contained a '/lib/firmware' directory while, if we say to the kernel build infrastructure to skip 'make firmware_install', there won't be any firmware directory at all in the resulting tree.

So once i fixed the kernel plugin by skipping the 'make firmware_install' step for the 'kernel-with-firmware: false' case, that test was suddenly failing:

--- a/snapcraft/plugins/kernel.py
+++ b/snapcraft/plugins/kernel.py
@@ -475,6 +475,8 @@ class KernelPlugin(kbuild.KBuildPlugin):
         # but snapd expects modules/ and firmware/
         shutil.move(
             os.path.join(self.installdir, 'lib', 'modules'), self.installdir)
-        shutil.move(
-            os.path.join(self.installdir, 'lib', 'firmware'), self.installdir)
+        if self.options.kernel_with_firmware:
+            shutil.move(
+                os.path.join(self.installdir, 'lib', 'firmware'),
+                self.installdir)
         os.rmdir(os.path.join(self.installdir, 'lib'))

It was os.rmdir() that was failing - we passed a firmware directory to the kernel plugin, but we didn't move it to the installdir since 'self.options.kernel_with_firmware == False' and when os.rmdir(lib) was invoked it found that dir to be non empty.

Reply:
Ah ha! I missed the rmdir, that makes more sense, thank you.